### ISSUE TITLE
Support for Exponential Constraints in Optimization Problems？

### ISSUE THREAD

**Issue by JianqiangDing** [oxfordcontrol/SOSTOOLS#18]
Hello,

I am working on an optimization problem where my objective function is a variable `t`. I am looking to set up constraints that involve the exponential of `t`, specifically constraints that include `exp(t)`.

Upon reviewing the code and documentation, I haven't found any mention of performing an exponential function operation on a `pvar` variable. Here is a brief overview of the problem I am trying to solve:

Objective:
minimize: -t

Constraints:
exp(t) <= some_constant

other linear matrix inequality constraints involving t

Could you please confirm if it is currently possible to define constraints that include `exp(t)`? If so, could you provide guidance or point me to the relevant documentation or examples that illustrate how to implement this?

Thank you for your assistance.

**Comment by djagt** [comment]
Hey,

SOS programming really only allows for polynomial constraints to be enforced, so constraints involving e.g. exponential functions are not supported. Accordingly, there is no exponential function for `pvar` objects. In some cases, you can still enforce the desired constraint as a polynomial one using a suitable change of variables, e.g. express the constraints in terms of s=exp(t) instead of t. However, in your case, is the "some_constant" fixed and known? If so, you can enforce the exponential constraint
```exp(t)<= c```
as a linear constraint
```t <= log(c)```
If all the other constraints are LMI constraints on `t`, you should be able to declare the problem without the need for exponential functions of your variables.

I hope this helps, but let me know if something is still unclear, or if you have any other questions/issues.

Declan

**Comment by JianqiangDing** [comment]
Hi, Declan.

Thank you very much for your prompt reply, your clear explanation did make me realise that without the variable replacement you hinted at, my problem was not directly applicable to the SOS framework.

My original problem was to optimise the determinant of a matrix via SOSTOOLS, after transform the objective `det(X)` to `log(det(X))` and by applying Schur's complement theorem, I think the problem can be converted to one that contains the following constraints (other constraints are all about X without t, like X is positive definite):

$$ \begin{bmatrix} X & I \\\ I & e^t I \end{bmatrix} \succeq 0$$

In this problem, the objective function is `min -t`, where `t` refers to the lower bound of `log(det(X))` such that `t<=log(det(X))` and I want to maximise the determinant of the matrix X. The other part of the constraint is to require the positive definiteness of X. In this case, t is optimised here as the objective function, and I'm not quite sure if change the variables as you propose will lead to an unequal problem?

this problem is original comes from [the document of Yalmip](https://yalmip.github.io/tutorial/maxdetprogramming/), I want to know how I can use SOSTOOLS to reproduce the code.

Thank you very much for helping me on this issue.

**Comment by djagt** [comment]
Sorry for the late response.

So if I understand correctly, your program is of the form

$$\min -t$$

$$\begin{bmatrix} X & I \\\ I& e^{t}I \end{bmatrix}\succeq 0$$

plus some other LMI constraints on just the matrix $X$ (not the variable $t$)?
In that case, I think you should be fine with the variable substitution $s=e^{t}$, minimizing $-s\leq 0$, and expressing the constraint in terms of just the variable $s\geq 0$. However I'm not sure I fully understand what you mean with a possible "unequal problem"?

Let me know if this doesn't answer your question.

Declan

**Comment by JianqiangDing** [comment]
Hi, Declan,

I sincerely apologize for not replying sooner. I've been deeply involved in revising a research paper, which has taken up most of my time these days.

I apologise that I did not explain my doubts very clearly before, my doubts about the transformation of the problem were mainly due to the fact that I considered the transformation of the problem using Schur's complementary theorem and then proceeded to replace the variables according to the idea that you proposed, so I subsequently found that the program written according to the idea did not give the optimal result, therefore, I have reconsidered the information again today and learnt about the methodology for Log-determinant mentioned in the [Mosek document](https://docs.mosek.com/modeling-cookbook/sdo.html#lm-complement-schur). The approach mentioned in Mosek's document, therefore, my simple use of Schur's theorem does not guarantee that the logdet optimisation problem will be transformed into another equivalent problem, thus reducing the solution space and ultimately not obtaining the desired minimum enclosing ellipse. Your idea is a very useful trick in convex optimisation problems and I am very grateful for your advice on my problem.

Additionally, I also check that how SOSTOOLS support logdet optimization, and I just find that the support for this kind of objective is mainly through sdpt3 solver, and the modification for the objective including logdet terms may be not too much intuitive in high dimensional space, I hope your team will refine the modifcaiton for this part in the furture.

Thank you very much for providing such a great tool and your kindly help, and again, I apologise to you for my delayed response. Thank you for your time and advice.
